### ISSUE TITLE
Name should not have prefix 'tower0' in this trainer

### ISSUE THREAD

**Issue by zxpeter** [tensorpack/tensorpack#1181]
Hi,
I'm using examples/fastercnn and try to add a self-implemented crf layer after mask-branch in model_mrcnn.py.

model_mrcnn.py
```
def maskrcnn_upXconv_head(feature, image, num_category, num_convs, norm=None):
    assert norm in [None, 'GN'], norm
    l = feature
    with argscope([Conv2D, Conv2DTranspose], data_format='channels_first',
                  kernel_initializer=tf.variance_scaling_initializer(
                      scale=2.0, mode='fan_out',
                      distribution='untruncated_normal' if get_tf_version_tuple() >= (1, 12) else 'normal')):
        for k in range(num_convs):
            l = Conv2D('fcn{}'.format(k), l, cfg.MRCNN.HEAD_DIM, 3, activation=tf.nn.relu)
            if norm is not None:
                l = GroupNorm('gn{}'.format(k), l)
        l = Conv2DTranspose('deconv', l, cfg.MRCNN.HEAD_DIM, 2, strides=2, activation=tf.nn.relu)
        l = Conv2D('conv', l, num_category, 1)
        l = tf.transpose(l, [0, 2, 3, 1]) 
        #add crf here
        l = CrfRnnLayer(image_dims=(28, 28),
                         num_classes=num_category,
                         theta_alpha=160.,
                         theta_beta=3.,
                         theta_gamma=3.,
                         num_iterations=10,
                         name='crfrnn')([l, image])
        l = tf.transpose(l, [0, 3, 1, 2]) 
    return l
```

crfrnn_layer.py
```
import keras.backend as K
from tensorpack.tfutils.scope_utils import under_name_scope
from tensorpack.models import layer_register
import numpy as np
import tensorflow as tf
from keras.engine.topology import Layer
import high_dim_filter_loader
custom_module = high_dim_filter_loader.custom_module

def _diagonal_initializer(shape):
    return np.eye(shape[0], shape[1], dtype=np.float32)
def _potts_model_initializer(shape):
    return -1 * _diagonal_initializer(shape)

class CrfRnnLayer(Layer):
    """ Implements the CRF-RNN layer described in:
    Conditional Random Fields as Recurrent Neural Networks,
    S. Zheng, S. Jayasumana, B. Romera-Paredes, V. Vineet, Z. Su, D. Du, C. Huang and P. Torr,
    ICCV 2015
    """
    def __init__(self, image_dims, num_classes,
                 theta_alpha, theta_beta, theta_gamma,
                 num_iterations, **kwargs):
        self.image_dims = image_dims
        self.num_classes = num_classes
        self.theta_alpha = theta_alpha
        self.theta_beta = theta_beta
        self.theta_gamma = theta_gamma
        self.num_iterations = num_iterations
        self.spatial_ker_weights = None
        self.bilateral_ker_weights = None
        self.compatibility_matrix = None
        super(CrfRnnLayer, self).__init__(**kwargs)

    def build(self, input_shape):
        # Weights of the spatial kernel
        self.spatial_ker_weights = self.add_weight(name='spatial_ker_weights',
                                                   shape=(self.num_classes, self.num_classes),
                                                   initializer=_diagonal_initializer,
                                                   trainable=True)

        # Weights of the bilateral kernel
        self.bilateral_ker_weights = self.add_weight(name='bilateral_ker_weights',
                                                     shape=(self.num_classes, self.num_classes),
                                                     initializer=_diagonal_initializer,
                                                     trainable=True)

        # Compatibility matrix
        self.compatibility_matrix = self.add_weight(name='compatibility_matrix',
                                                    shape=(self.num_classes, self.num_classes),
                                                    initializer=_potts_model_initializer,
                                                    trainable=True)

        super(CrfRnnLayer, self).build(input_shape)

    def call(self, inputs):
        unaries = tf.transpose(inputs[0][0, :, :, :], perm=(2, 0, 1))
        rgb = tf.transpose(inputs[1][0, :, :, :], perm=(2, 0, 1))

        c, h, w = self.num_classes, self.image_dims[0], self.image_dims[1]
        all_ones = np.ones((c, h, w), dtype=np.float32)

        # Prepare filter normalization coefficients
        spatial_norm_vals = custom_module.high_dim_filter(all_ones, rgb, bilateral=False,
                                                          theta_gamma=self.theta_gamma)
        bilateral_norm_vals = custom_module.high_dim_filter(all_ones, rgb, bilateral=True,
                                                            theta_alpha=self.theta_alpha,
                                                            theta_beta=self.theta_beta)
        q_values = unaries
        for i in range(self.num_iterations):
            softmax_out = tf.nn.softmax(q_values, 0)
            # Spatial filtering
            spatial_out = custom_module.high_dim_filter(softmax_out, rgb, bilateral=False,
                                                        theta_gamma=self.theta_gamma)
            # spatial_out = K.print_tensor(spatial_out, message='upscore = ')
            spatial_out = spatial_out / spatial_norm_vals

            # Bilateral filtering
            bilateral_out = custom_module.high_dim_filter(softmax_out, rgb, bilateral=True,
                                                          theta_alpha=self.theta_alpha,
                                                          theta_beta=self.theta_beta)
            bilateral_out = bilateral_out / bilateral_norm_vals

            # Weighting filter outputs
            message_passing = (tf.matmul(self.spatial_ker_weights,
                                         tf.reshape(spatial_out, (c, -1))) +
                               tf.matmul(self.bilateral_ker_weights,
                                         tf.reshape(bilateral_out, (c, -1))))
            # Compatibility transform
            pairwise = tf.matmul(self.compatibility_matrix, message_passing)
            # Adding unary potentials
            pairwise = tf.reshape(pairwise, (c, h, w))
            q_values = unaries - pairwise
        return tf.transpose(tf.reshape(q_values, (1, c, h, w)), perm=(0, 2, 3, 1))
    def compute_output_shape(self, input_shape):
        return input_shape
```

And I got the error message:
```
Do not know how to sync variable 'tower0/maskrcnn/crfrnn/spatial_ker_weights:0' across GPUs. 
Reason: Name should not have prefix 'tower0' in this trainer! 
Traceback (most recent call last):
  File "./train.py", line 592, in <module>
    launch_train_with_config(traincfg, trainer)
  File "/home/user03/anaconda3/lib/python3.6/site-packages/tensorpack/train/interface.py", line 91, in launch_train_with_config
    model._build_graph_get_cost, model.get_optimizer)
  File "/home/user03/anaconda3/lib/python3.6/site-packages/tensorpack/utils/argtools.py", line 176, in wrapper
    return func(*args, **kwargs)
  File "/home/user03/anaconda3/lib/python3.6/site-packages/tensorpack/train/tower.py", line 219, in setup_graph
    train_callbacks = self._setup_graph(input, get_cost_fn, get_opt_fn)
  File "/home/user03/anaconda3/lib/python3.6/site-packages/tensorpack/train/trainers.py", line 190, in _setup_graph
    self.train_op, post_init_op = self._builder.build(grad_list, get_opt_fn)
  File "/home/user03/anaconda3/lib/python3.6/site-packages/tensorpack/graph_builder/training.py", line 303, in build
    post_init_op = SyncMultiGPUReplicatedBuilder.get_post_init_ops()
  File "/home/user03/anaconda3/lib/python3.6/site-packages/tensorpack/graph_builder/training.py", line 332, in get_post_init_ops
    log_failure(v.name, "Name should not have prefix 'tower0' in this trainer!")
  File "/home/user03/anaconda3/lib/python3.6/site-packages/tensorpack/graph_builder/training.py", line 322, in log_failure
    "The aforementioned variable is trainable, so this is probably a fatal error."
AssertionError: The aforementioned variable is trainable, so this is probably a fatal error.
```
I also checked related issues #325 and #1153.
The crfrnn_layer.py is implemented with keras and I think maybe the problem is below as you mentioned in #325.
> I believe that's because Keras use tf.Variable to create variables instead of tf.get_variable.

How can I get rid of the prefix 'tower0' in the trainer if I really want use crfrnn_layer.py?
Where should I add the @under_name_scope() and @layer_register(log_shape=True)?
Also how to implement a new layer using tensorpack rather than keras?

**Comment by ppwwyyxx** [comment]
Keras layers do not respect variable scope, which fundamentally conflicts with the design of tensorpack multigpu trainer. Your usage will work if you use a single-GPU trainer (`SimpleTrainer`).
Multi gpu training with Keras layers is doable if your whole model is in Keras (https://github.com/tensorpack/tensorpack/tree/master/examples/keras). But these are all experimental since Keras fundamentally conflicts with tensorflow scoping.

**Comment by ppwwyyxx** [comment]
I found that it's not hard to make "SyncMultiGPUTrainerParameterServer" to suport using Keras inside tensorpack model. See https://github.com/tensorpack/tensorpack/blob/6926d22a88547899e390ebe709913c38382ff361/examples/keras/mnist-keras.py#L27-L41 for a working example. Note that you still cannot create the Keras model inside the tensorpack function: it has to be a separate memoized function as the above example shows.
Supporting the more efficient replicated trainer will need more hacks due to how Keras works (and is one of the main reasons why Keras's multi gpu trainer is very slow).

Closing since Keras is not something we'd like to support anyway.
